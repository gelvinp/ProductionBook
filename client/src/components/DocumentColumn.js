import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Segment, Loader, Input, Dimmer, Icon, Button } from 'semantic-ui-react'
import SectionList from '../containers/SectionListContainer.js'
import UploadModal from '../containers/UploadModalContainer.js'
import SectionModal from '../containers/SectionModalContainer.js'
import DocumentModal from '../containers/DocumentModalContainer.js'

class DocumentColumn extends Component {
  state = {
    sectionField: '',
    sectionError: false,
  }
  handleKeyDown = (e, func) => {
    if (e.key === 'Enter') {
      func()
    }
  }
  handleCreateSection = async () => {
    const { sectionField } = this.state
    const { submitSection, createSection } = this.props
    submitSection(sectionField).then(json => {
      if (json.error || json.data.id === -1) {
        this.setState({ sectionError: true })
      } else {
        createSection(json.data.id, sectionField)
        this.setState({ sectionField: '', sectionError: false })
      }
    })
  }
  render() {
    const { openUpload, loading, mobile } = this.props
    const { sectionField, sectionError } = this.state
    return (
      <Dimmer.Dimmable
        as={Segment}
        style={{
          height: '100%',
          width: '100%',
          display: 'flex',
          flexDirection: 'column',
          overflow: 'hidden',
          ...(mobile ? { padding: '1em', marginTop: 0 } : {}),
        }}
      >
        <SectionList mobile={mobile} />
        <Button
          id="openUploadButton"
          onClick={openUpload}
          basic
          icon
          labelPosition="left"
          style={{
            marginBottom: '0.5rem',
            width: '12rem',
            maxHeight: '6rem',
          }}
        >
          <Icon name="upload" />
          Upload File
        </Button>
        <Input
          value={sectionField}
          id="sectionFieldInput"
          onChange={e => this.setState({ sectionField: e.target.value })}
          placeholder="Add a section"
          onKeyDown={e => this.handleKeyDown(e, this.handleCreateSection)}
          error={sectionError}
          label={
            <Button
              icon="plus"
              id="submitSectionButton"
              basic
              onClick={this.handleCreateSection}
              disabled={!sectionField.length}
            />
          }
          labelPosition="right"
        />
        <UploadModal />
        <SectionModal />
        <DocumentModal />
        <Dimmer active={loading}>
          <Loader>Loading</Loader>
        </Dimmer>
      </Dimmer.Dimmable>
    )
  }
}

DocumentColumn.propTypes = {
  openUpload: PropTypes.func.isRequired,
  createSection: PropTypes.func.isRequired,
  submitSection: PropTypes.func.isRequired,
  mobile: PropTypes.bool,
  loading: PropTypes.bool.isRequired,
}

export default DocumentColumn
