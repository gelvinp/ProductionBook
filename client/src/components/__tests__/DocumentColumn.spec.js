import React from 'react'
import DocumentColumn from '../DocumentColumn.js'
import { shallow } from 'enzyme'

describe('DocumentColumn', () => {
  it('Displays correctly', () => {
    const wrapper = shallow(
      <DocumentColumn
        openUpload={jest.fn()}
        createSection={jest.fn()}
        submitSection={jest.fn()}
        loading={false}
      />
    )
    expect(wrapper).toMatchSnapshot()
  })

  it('Enables the section submit button when text is entered', () => {
    const wrapper = shallow(
      <DocumentColumn
        openUpload={jest.fn()}
        createSection={jest.fn()}
        submitSection={jest.fn()}
        loading={false}
      />
    )
    expect(
      wrapper.find('#sectionFieldInput').props().label.props.disabled
    ).toBeTruthy()
    wrapper.setState({ sectionField: 'a' })
    expect(
      wrapper.find('#sectionFieldInput').props().label.props.disabled
    ).not.toBeTruthy()
  })

  it('Updates the sectionField', () => {
    const wrapper = shallow(
      <DocumentColumn
        openUpload={jest.fn()}
        createSection={jest.fn()}
        submitSection={jest.fn()}
        loading={false}
      />
    )
    wrapper
      .find('#sectionFieldInput')
      .simulate('change', { target: { value: 'Test' } })
    expect(wrapper.state('sectionField')).toEqual('Test')
  })

  it('Opens the upload modal', () => {
    const openUpload = jest.fn()
    const wrapper = shallow(
      <DocumentColumn
        openUpload={openUpload}
        createSection={jest.fn()}
        submitSection={jest.fn()}
        loading={false}
      />
    )
    wrapper.find('#openUploadButton').simulate('click')
    expect(openUpload).toHaveBeenCalled()
  })

  it('Handles a communication error', async () => {
    expect.assertions(3)
    const submitSection = jest.fn().mockImplementation(section => {
      expect(section).toEqual('test')
      return Promise.resolve({
        error: true,
      })
    })
    const wrapper = shallow(
      <DocumentColumn
        openUpload={jest.fn()}
        submitSection={submitSection}
        loading={false}
        createSection={jest.fn()}
      />
    )
    wrapper.setState({ sectionField: 'test' })
    await wrapper
      .find('#sectionFieldInput')
      .props()
      .label.props.onClick()
    expect(wrapper.state('sectionError')).toBeTruthy()
    expect(wrapper.find('#sectionFieldInput').props().error).toBeTruthy()
  })

  it('Handles an invalid section', async () => {
    expect.assertions(3)
    const submitSection = jest.fn().mockImplementation(section => {
      expect(section).toEqual('test')
      return Promise.resolve({
        error: false,
        data: {
          id: -1,
        },
      })
    })
    const wrapper = shallow(
      <DocumentColumn
        openUpload={jest.fn()}
        submitSection={submitSection}
        loading={false}
        createSection={jest.fn()}
      />
    )
    wrapper.setState({ sectionField: 'test' })
    await wrapper
      .find('#sectionFieldInput')
      .props()
      .label.props.onClick()
    expect(wrapper.state('sectionError')).toBeTruthy()
    expect(wrapper.find('#sectionFieldInput').props().error).toBeTruthy()
  })

  it('Handles a valid section', async () => {
    expect.assertions(6)
    const submitSection = jest.fn().mockImplementation(section => {
      expect(section).toEqual('test')
      return Promise.resolve({
        error: false,
        data: {
          id: 1,
        },
      })
    })
    const createSection = jest.fn().mockImplementation((id, name) => {
      expect(id).toBe(1)
      expect(name).toEqual('test')
    })
    const wrapper = shallow(
      <DocumentColumn
        openUpload={jest.fn()}
        submitSection={submitSection}
        loading={false}
        createSection={createSection}
      />
    )
    wrapper.setState({ sectionField: 'test' })
    await wrapper
      .find('#sectionFieldInput')
      .props()
      .label.props.onClick()
    expect(wrapper.state('sectionError')).not.toBeTruthy()
    expect(wrapper.state('sectionField')).toEqual('')
    expect(wrapper.find('#sectionFieldInput').props().error).not.toBeTruthy()
  })

  it('Rejects non enter keydown', () => {
    const submitSection = jest.fn().mockImplementation(section => {
      return Promise.resolve({
        error: true,
      })
    })
    const wrapper = shallow(
      <DocumentColumn
        openUpload={jest.fn()}
        submitSection={submitSection}
        loading={false}
        createSection={jest.fn()}
      />
    )
    wrapper.find('#sectionFieldInput').simulate('keydown', { key: 'Down' })
    expect(submitSection).not.toHaveBeenCalled()
  })

  it('Accepts enter keydown', () => {
    const submitSection = jest.fn().mockImplementation(section => {
      return Promise.resolve({
        error: true,
      })
    })
    const wrapper = shallow(
      <DocumentColumn
        openUpload={jest.fn()}
        submitSection={submitSection}
        loading={false}
        createSection={jest.fn()}
      />
    )
    wrapper.find('#sectionFieldInput').simulate('keydown', { key: 'Enter' })
    expect(submitSection).toHaveBeenCalled()
  })

  it('Does not have padding and top margin on desktop', () => {
    const wrapper = shallow(
      <DocumentColumn
        openUpload={jest.fn()}
        createSection={jest.fn()}
        submitSection={jest.fn()}
        loading={false}
      />
    )
    expect(wrapper.find('DimmerDimmable').props().style).not.toHaveProperty(
      'padding'
    )
    expect(wrapper.find('DimmerDimmable').props().style).not.toHaveProperty(
      'marginTop'
    )
  })

  it('Does have padding and top margin on mobile', () => {
    const wrapper = shallow(
      <DocumentColumn
        openUpload={jest.fn()}
        createSection={jest.fn()}
        submitSection={jest.fn()}
        loading={false}
        mobile
      />
    )
    expect(wrapper.find('DimmerDimmable').props().style).toHaveProperty(
      'padding',
      '1em'
    )
    expect(wrapper.find('DimmerDimmable').props().style).toHaveProperty(
      'marginTop'
    )
  })
})
