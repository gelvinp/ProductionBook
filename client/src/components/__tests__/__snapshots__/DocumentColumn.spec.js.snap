// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentColumn Displays correctly 1`] = `
<DimmerDimmable
  as={[Function]}
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
      "height": "100%",
      "overflow": "hidden",
      "width": "100%",
    }
  }
>
  <Connect(SectionList) />
  <Button
    as="button"
    basic={true}
    icon={true}
    id="openUploadButton"
    labelPosition="left"
    onClick={[MockFunction]}
    style={
      Object {
        "marginBottom": "0.5rem",
        "maxHeight": "6rem",
        "width": "12rem",
      }
    }
  >
    <Icon
      as="i"
      name="upload"
    />
    Upload File
  </Button>
  <Input
    error={false}
    id="sectionFieldInput"
    label={
      <Button
        as="button"
        basic={true}
        disabled={true}
        icon="plus"
        id="submitSectionButton"
        onClick={[Function]}
      />
    }
    labelPosition="right"
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="Add a section"
    type="text"
    value=""
  />
  <Connect(UploadModal) />
  <Connect(SectionModal) />
  <Connect(DocumentModal) />
  <Dimmer
    active={false}
  >
    <Loader>
      Loading
    </Loader>
  </Dimmer>
</DimmerDimmable>
`;
